Project feature warning system

aaronfranke opened the pull request:
(Note: This PR also increments the config version by necessity)

A mostly complete implementation of a feature warning system, implements and closes https://github.com/godotengine/godot-proposals/issues/427, implements and closes https://github.com/godotengine/godot-proposals/issues/237

EDIT: The screenshot is a bit outdated, since now there are more specific error messages available.

![f](https://user-images.githubusercontent.com/1646875/62596124-cec27880-b895-11e9-883d-ad3d6e4e5da5.png)

What works: Reading project features from `project.godot`, displaying unsupported features, warning when trying to open a project with unsupported features, changing the features to comply with currently supported and required features, and saving project features to `project.godot`.

This PR uses the new feature system for the following:

* Saves the Godot version. This is in parallel to `config_version`, so as an example if 4.0 and 4.1 use the same config version then this system will show a warning when opening 4.0 projects in 4.1 and 4.1 projects in 4.0, but if 4.2 uses a higher config version then upgrading to 4.2 would be one-way.
  * Users can also pin a project to a minor version or engine build if they wish, so 4.0.1 as an example.
* Allows different builds of Godot to save feature tags to warn about the different features. In this PR it saves information about:
  * Whether a project is using single-precision or doubles.
  * Which renderer the project is using (Vulkan/GLES/etc).
  * Whether a project is using C#.
  * More can be added later if there's a need for any other feature tags.
* Allows people to make custom builds of Godot and include a feature tag for it. Since this system just saves strings, even official builds of Godot can warn that they are missing features found in custom Godot forks.

groud commented:
It looks globally good to me, though the way things are implemented are kind of spread around.
I would have preferred all the logic to be in the same place if possible, so that's it's easier to modify if we add new features to be considered.

Calinou commented on core/config/project_settings.h:
```diff
@@ -108,6 +110,10 @@ class ProjectSettings : public Object {
 
 	Error _save_custom_bnd(const String &p_file);
 
+	PackedStringArray _get_supported_features();
+	PackedStringArray _get_unsupported_features(PackedStringArray &p_project_features);
+	PackedStringArray _trim_to_supported_features(PackedStringArray &p_project_features);
```
I think these 3 methods could be made `const`, double-check that.

aaronfranke replied:
Done, I added `const` everywhere it was relevant in this PR.

Calinou commented on core/config/project_settings.cpp:
```diff
@@ -56,6 +58,79 @@ String ProjectSettings::get_resource_path() const {
+	PackedStringArray features = PackedStringArray();
+	features.append(VERSION_BRANCH);
+#ifdef REAL_T_IS_DOUBLE
+	features.append("Double");
+#else
+	features.append("Single");
```
If there's enough space, I think it'd be better to be explicit that this is about precision:

```suggestion
	features.append("DoublePrecision");
#else
	features.append("SinglePrecision");
```

Alternatively, this could be `DoublePhysics` and `SinglePhysics` respectively, but this could be confused for a physics engine.

aaronfranke replied:
Changed to end in `-precision`. Testing this also made me discover a bug caused by #45968 as now I have to explicitly call `save();` when updating the project features, so that's fixed now :)

Calinou commented on editor/project_manager.cpp:
```diff
@@ -2102,6 +2180,13 @@ void ProjectManager::_open_selected_projects_ask() {
 		return;
 	}
+	// Check if the project is using features not supported by this build of Godot.
+	if (unsupported_features.length() > 1) {
```
Shouldn't this be `unsupported_features.length() >= 1`?

aaronfranke replied:
Yes, I must have forgotten to test 1-character features. Fixed.

Calinou reviewed (approved):
I haven't been able to test this locally, but this looks good to me :+1:

KoBeWi commented:
Any advice how to test this easily?

aaronfranke commented:
@KoBeWi Basically, test that it does what it's intended to do:

* When creating a new project, it should create a list of features.
* When opening an old project, it should upgrade the project config version and generate a list of features.
* If something in the project's feature list is not something supported:
  * The project manager should warn about this, showing both yellow text, and a warning when clicked on.
  * If the project is opened, the unsupported features should be trimmed away, and any required ones are added.
* If Mono is enabled, the C# feature is marked as supported, and the project manager won't warn about it.
* The C# feature's existence in the list of the project's features depends on the existence of a `.csproj` file.

To quickly test how unsupported features behave, you can manually edit the `config/features=PackedStringArray` line in the `[application]` section of `project.godot`.

KoBeWi commented:
When a project has `config_version` of 5, but no feature list, you get this crash:
```
CrashHandlerException: Program crashed
Dumping the backtrace.
[0] Object::get (C:\godot_source\core\object\object.cpp:445)
[1] Control::is_layout_rtl (C:\godot_source\scene\gui\control.cpp:442)
[2] BoxContainer::_resort (C:\godot_source\scene\gui\box_container.cpp:47)
[3] BoxContainer::set_alignment (C:\godot_source\scene\gui\box_container.cpp:310)
[4] ProjectList::create_project_item_control (C:\godot_source\editor\project_manager.cpp:1385)
[5] ProjectList::load_projects (C:\godot_source\editor\project_manager.cpp:1303)
[6] ProjectManager::_load_recent_projects (C:\godot_source\editor\project_manager.cpp:2052)
[7] ProjectManager::ProjectManager (C:\godot_source\editor\project_manager.cpp:2778)
[8] Main::start (C:\godot_source\main\main.cpp:2359)
[9] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:161)
[10] _main (C:\godot_source\platform\windows\godot_windows.cpp:185)
[11] main (C:\godot_source\platform\windows\godot_windows.cpp:199)
[12] __scrt_common_main_seh (d:\agent\_work\63\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[13] BaseThreadInitThunk
-- END OF BACKTRACE --
```
It can happen when you just open and close the old project. The feature list won't appear, which will cause a crash on next project manager run (and will automatically add features).

IMO `config_version=5` and no features, even if theoretically should be impossible, it shouldn't cause crash.
Also Project Manager modifying `project.godot` before the user does anything is a bit shady.

aaronfranke commented:
@KoBeWi See this comment: https://github.com/godotengine/godot/pull/41100#issuecomment-745226783

This crash did not exist when I first created this PR, and really the code in `Control::is_layout_rtl` should be fixed. Maybe there is something that I can do to workaround this issue in this PR, if so I'm not sure what it is, because the stack trace is not directly related to any of the lines of code changed in this PR.

KoBeWi commented:
I think the crash is related to this PR. It only happens when `config/features` doesn't exist in a project with `config_version` 5. I modified `is_layout_rtl()` to always return false and I get a crash in another place.

KoBeWi commented on core/config/project_settings.cpp:
```diff
@@ -56,6 +58,85 @@ String ProjectSettings::get_resource_path() const {
+
+// Returns the features that this project needs but this build of Godot lacks, as a user-friendly string.
+// This is used by the project manager to display and warn about unsupported features.
+const String ProjectSettings::get_unsupported_features_str(const PackedStringArray &p_project_features) {
+	PackedStringArray unsupported_features = singleton->_get_unsupported_features(p_project_features);
+	String features_string = Variant(unsupported_features).operator String();
+	if (features_string.length() < 3) {
+		return "";
+	}
+	return features_string.trim_prefix("[").trim_suffix("]");
+}
```
If this is going to be used only by Project Manager, I'd move this method to project manager and make `_get_unsupported_features()` public.

KoBeWi commented:
Another thing I noticed is that creating new project adds `Vulkan` feature, but opening a project doesn't. This should be probably made consistent.

aaronfranke commented:
@KoBeWi For that one, see https://github.com/godotengine/godot/issues/37645 It's not possible to tell with absolute certainty what the API should be when opening a project if it's not explicitly saved. For example, what if the project was created in another version of Godot where the default isn't Vulkan?

KoBeWi commented:
Well then this could be saved by EditorNode. The ProjectSettings save upon opening the project (see #47385). At this point it should be possible to determine the used API I think. But then the project features wouldn't be nicely exclusive to project manager, so not sure 🤔

KoBeWi commented on editor/project_manager.cpp:
```diff
@@ -1174,13 +1182,57 @@ void ProjectList::load_project_data(const String &p_property_key, Item &p_item,
+		project_features_dirty = true;
+	}
+	// Check the rendering API.
+	const String rendering_api = cf->get_value("rendering", "quality/driver/driver_name", "Unknown");
+	if (rendering_api != "Unknown") {
```
This could as well be an empty string instead of Unknown.

KoBeWi commented on editor/project_manager.cpp:
```diff
@@ -1174,13 +1182,57 @@ void ProjectList::load_project_data(const String &p_property_key, Item &p_item,
+	}
+	// Write the updated feature list, but only if the project config version is the same as ours.
+	// We should never write to project files with a different config version.
+	if (project_features_dirty && config_version == ProjectSettings::CONFIG_VERSION) {
+		ProjectSettings *ps = memnew(ProjectSettings);
+		ps->load_custom(conf);
+		ps->set("application/config/features", project_features);
+		ps->save_custom(conf);
+		memdelete(ps);
+	}
```
I noticed that you do `sort()` and set `project_features_dirty` to true each time. The sorting could just be done in this block, single time. Or in another condition block that only checks `project_features_dirty`.

KoBeWi reviewed (changes requested):
Ok I went through the whole code. The feature is good and the implementation is also ok. I left a few nitpicks. The Vulkan and `_get_unsupported_features()` comments are not very important, could be done in another PR probably.

The crash needs to be fixed (unless you can somehow prove it's not related to your changes). Not sure what could be causing it though.

aaronfranke commented:
* Changed the rendering driver string to `""` instead of `"Unknown"`.
* Fixed sorting potentially uselessly happening multiple times.
* Fixed the crash. It happened due to changes to `ProjectSettings` in master such that the way this PR was using it caused a crash, however I think that the way this PR previously used `ProjectSettings` was never valid to begin with when this was based on 3.x even though it didn't crash there (it was creating an instance while it's supposed to be a singleton).

Calinou commented on core/config/project_settings.cpp:
```diff
@@ -54,6 +56,73 @@ String ProjectSettings::get_resource_path() const {
+	features.append(VERSION_FULL_BUILD);
+	// For now, assume Vulkan is always supported.
+	// This should be removed if it's possible to build the editor without Vulkan.
+	features.append("Vulkan 1.0 Clustered");
+	features.append("Vulkan 1.0 Mobile");
```
It's possible for Godot to use extensions from Vulkan 1.1 and 1.2 if supported. Godot will do so automatically as needed (e.g. for multiview). Should the version 1.0 still be listed in this case?

aaronfranke replied:
This is a good question, which I'm not sure of the answer to. I included the minor version just in case, but if Vulkan extensions are used automatically yet not required, perhaps this isn't necessary. Is there a chance that Godot would require Vulkan 1.x versions in the future? If so, "1.0" makes sense, if not, it should be "Vulkan 1" (and not just "Vulkan" in the event that Vulkan 2+ is significantly different).

Calinou replied:
> Is there a chance that Godot would require Vulkan 1.x versions in the future?

I don't know for sure, but it's unlikely because Vulkan 1.1 support is still poor on mobile platforms right now.

Calinou commented on editor/project_manager.cpp:
```diff
@@ -1364,40 +1428,64 @@ void ProjectList::create_project_item_control(int p_index) {
+			Label *unsupported_label = memnew(Label(item.unsupported_features_str));
+			unsupported_label->set_custom_minimum_size(Size2(length * 15, 10));
+			unsupported_label->add_theme_font_override("font", get_theme_font(SNAME("title"), SNAME("EditorFonts")));
+			unsupported_label->add_theme_color_override("font_color", Color(1, 0.85, 0.4));
```
`warning_color` should be sourced from the editor theme so that the color will be adjusted on light themes automatically.

aaronfranke replied:
When I do `get_theme_color(SNAME("warning_color"))` it returns black.

KoBeWi replied:
Use `get_theme_color(SNAME("warning_color"), SNAME("Editor"))`

Calinou commented on editor/project_manager.cpp:
```diff
@@ -473,8 +473,20 @@ class ProjectDialog : public ConfirmationDialog {
+					int renderer_type = rasterizer_button_group->get_pressed_button()->get_meta(SNAME("driver_name"));
+					initial_settings["rendering/vulkan/rendering/back_end"] = renderer_type;
+					if (renderer_type == 0) {
+						project_features.push_back("Vulkan 1.0 Clustered");
+					} else if (renderer_type == 1) {
+						project_features.push_back("Vulkan 1.0 Mobile");
```
Remember to also update this part about Vulkan versioning if needed.

aaronfranke commented:
@reduz Is that guaranteed to happen for 4.0?

Anyway, this PR solves more problems than just that. It also:

* Saves the Godot version. This is in parallel to `config_version`, so as an example if 4.0 and 4.1 use the same config version then this system will show a warning when opening 4.0 projects in 4.1 and 4.1 projects in 4.0, but if 4.2 uses a higher config version then upgrading to 4.2 would be one-way.
  * Users can also pin a project to a minor version or engine build if they wish, so 4.0.1 as an example.
* Allows different builds of Godot to save feature tags to warn about the different features. In this PR it saves information about:
  * Whether a project is using single-precision or doubles.
  * Which renderer the project is using (Vulkan/GLES/etc).
  * Whether a project is using C#.
* Allows people to make custom builds of Godot and include a feature tag for it. Since this system just saves strings, even official builds of Godot can warn that they are missing features found in custom Godot forks.

There are other ways to solve most of these problems, but the last point is only possible with a generic system that stores strings.

reduz commented:
> @reduz Is that guaranteed to happen for 4.0?

Yes that is the idea, otherwise it would be a major compatibility breakage between 4.0 and 4.1.

>   Saves the Godot version. This is in parallel to config_version, so as an example if 4.0 and 4.1 use the same config version then this system will show a warning when opening 4.0 projects in 4.1 and 4.1 projects in 4.0, but if 4.2 uses a higher config version then upgrading to 4.2 would be one-way.

Not against a version check, this part may go through, but needs to be consulted with other contributors.

For the rest, I think using feature tags for this is not a good idea, because most of them are either irrelevant or platform dependent. IMO something more specialized is required for this situation. I think this needs to be something else more related to compilation flags (engine version, whether it uses doubles, etc) and not feature tags.

I suggest going back to the drawing board with this and create a new proposal.

Calinou commented:
> I think this needs to be something else more related to compilation flags (engine version, whether it uses doubles, etc) and not feature tags.

All current feature tags are determined at compile-time, so double precision support could be a feature tag.

akien-mga commented on editor/project_manager.cpp:
```diff
@@ -473,8 +473,20 @@ class ProjectDialog : public ConfirmationDialog {
+					// Be sure to change this code if/when renderers are changed.
+					int renderer_type = rasterizer_button_group->get_pressed_button()->get_meta(SNAME("driver_name"));
+					initial_settings["rendering/vulkan/rendering/back_end"] = renderer_type;
+					if (renderer_type == 0) {
+						project_features.push_back("Vulkan 1 Clustered");
+					} else if (renderer_type == 1) {
+						project_features.push_back("Vulkan 1 Mobile");
+					} else {
+						WARN_PRINT("Unknown renderer type. Please report this as a bug on GitHub.");
```
This shouldn't be hardcoded (it's no longer accurate either).

aaronfranke replied:
Do you happen to know where the enum is?

aaronfranke replied:
Looks like it's hard-coded here too (`servers/rendering_server.cpp:2827`):

```
GLOBAL_DEF_RST_BASIC("rendering/vulkan/rendering/back_end", 0);
GLOBAL_DEF_RST_BASIC("rendering/vulkan/rendering/back_end.mobile", 1);
ProjectSettings::get_singleton()->set_custom_property_info("rendering/vulkan/rendering/back_end",
		PropertyInfo(Variant::INT,
				"rendering/vulkan/rendering/back_end",
				PROPERTY_HINT_ENUM, "Forward Clustered (Supports Desktop Only),Forward Mobile (Supports Desktop and Mobile)"));
```

akien-mga replied:
Hm right, I guess it's fine for now. We can rethink this once the OpenGL backend is ready and needs to be handled too.

akien-mga commented on editor/project_manager.cpp:
```diff
@@ -473,8 +473,20 @@ class ProjectDialog : public ConfirmationDialog {
+					int renderer_type = rasterizer_button_group->get_pressed_button()->get_meta(SNAME("driver_name"));
+					initial_settings["rendering/vulkan/rendering/back_end"] = renderer_type;
+					if (renderer_type == 0) {
+						project_features.push_back("Vulkan 1 Clustered");
+					} else if (renderer_type == 1) {
+						project_features.push_back("Vulkan 1 Mobile");
```
I wouldn't put the "1" as it doesn't really matter (for now).

aaronfranke replied:
I am somewhat concerned about what happens if we have a Vulkan 2.x renderer (whenever that comes into existence), but I guess then users will just know that no number means Vulkan 1.x.

akien-mga commented on core/config/project_settings.cpp:
```diff
@@ -66,6 +68,73 @@ String ProjectSettings::get_imported_files_path() const {
+	PackedStringArray features = PackedStringArray();
+	features.append(VERSION_BRANCH);
+#ifdef REAL_T_IS_DOUBLE
+	features.append("Double-precision");
```
Would need to settle on a format, "Double-precision" and "Vulkan Clustered" are not consistent.

aaronfranke replied:
I went with "Double Precision". Also I removed "Single-precision" since reduz said it wasn't necessary as a warning, which is mostly true.

akien-mga reviewed (approved):
Approved in PR meeting.

We might want to iterate on the actual implementation to see what makes the most sense, but this is a good start to build upon.

aaronfranke commented on editor/project_manager.cpp:
```diff
@@ -2135,6 +2220,35 @@ void ProjectManager::_open_selected_projects_ask() {
 	}
+	// Check if the project is using features not supported by this build of Godot.
+	if (!unsupported_features.is_empty()) {
+		String warning_message = "";
+		for (int i = 0; i < unsupported_features.size(); i++) {
+			String feature = unsupported_features[i];
+			if (feature == "Double Precision") {
+				warning_message += TTR("Warning: This project uses double precision floats, but this version of\nGodot uses single precision floats. Opening this project may cause data loss.\n\n");
+				unsupported_features.remove(i);
+				i--;
+			} else if (feature == "C#") {
+				warning_message += TTR("Warning: This project uses C#, but this build of Godot does not have\nthe Mono module. If you proceed you will not be able to use any C# scripts.\n\n");
+				unsupported_features.remove(i);
+				i--;
+			} else if (feature.substr(0, 3).is_numeric()) {
+				warning_message += vformat(TTR("Warning: This project was built in Godot %s.\nOpening will upgrade or downgrade the project to Godot %s.\n\n"), Variant(feature), Variant(VERSION_BRANCH));
+				unsupported_features.remove(i);
+				i--;
+			}
+		}
+		if (!unsupported_features.is_empty()) {
+			String unsupported_features_str = Variant(unsupported_features).operator String().trim_prefix("[").trim_suffix("]");
+			warning_message += vformat(TTR("Warning: This project uses the following features not supported by this build of Godot:\n\n%s\n\n"), unsupported_features_str);
+		}
+		warning_message += TTR("Open anyway? Project will be modified.");
```
This code has been updated to have more specific error messages, but still have a fallback for anything not covered by those messages.

akien-mga commented on core/config/project_settings.cpp:
```diff
@@ -41,6 +41,8 @@
 #include "core/os/keyboard.h"
 #include "core/os/os.h"
 #include "core/variant/variant_parser.h"
+#include "core/version.h"
+#include "modules/modules_enabled.gen.h"
```
To follow the same style as #54912:
```suggestion
#include "core/version.h"

#include "modules/modules_enabled.gen.h" // For mono.
```

aaronfranke committed:
Project feature warning system

akien-mga commented:
You can remove me as co-author, it's not deserved for a style nitpick ;)

akien-mga commented:
Thanks!
